package com.example.data.client.impl

import android.content.Context
import com.example.data.client.api.NetworkClient
import com.example.data.models.main.MainDataResponseDto
import com.example.data.models.search.RecommendTicketsDto
import com.example.data.models.tickets.TicketsDto
import com.example.network.MokApi
import com.example.util.isConnected
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import retrofit2.HttpException
import java.net.ConnectException
import java.net.SocketTimeoutException

class RetrofitNetworkClient(private val service: MokApi, private val context: Context) :
    NetworkClient {
    override suspend fun getMainData(): Result<MainDataResponseDto> {
        if (!isConnected(context)) {
            return Result.failure(ConnectException())
        }
        val response = withContext(Dispatchers.IO) {
            try {
                val result = service.getMainData()
                Result.success(result)
            } catch (e: HttpException) {
                Result.failure(e)
            } catch (e: SocketTimeoutException) {
                Result.failure(e)
            }
        }
        return response
    }

    override suspend fun getRecommendTickets(): Result<RecommendTicketsDto> {
        if (!isConnected(context)) {
            return Result.failure(ConnectException())
        }
        val response = withContext(Dispatchers.IO) {
            try {
                val result = service.getRecommendTickets()
                Result.success(result)
            } catch (e: HttpException) {
                Result.failure(e)
            } catch (e: SocketTimeoutException) {
                Result.failure(e)
            }
        }
        return response
    }

    override suspend fun getAllTickets(): Result<TicketsDto> {
        if (!isConnected(context)) {
            return Result.failure(ConnectException())
        }
        val response = withContext(Dispatchers.IO) {
            try {
                val result = service.getAllTickets()
                Result.success(result)
            } catch (e: HttpException) {
                Result.failure(e)
            } catch (e: SocketTimeoutException) {
                Result.failure(e)
            }
        }
        return response
    }
}